File: Leetcode_200. Number of Islands.cpp
/*
Runtime: 64 ms, faster than 35.22% of C++ online submissions for Number of Islands.
Memory Usage: 12.4 MB, less than 44.87% of C++ online submissions for Number of Islands.
*/


class Solution {
public:
    bool check[301][301] = {false};
    int dr[4] = {0,0,1,-1};
    int dc[4] = {1, -1, 0,0};
    
    void dfs(int r, int c, vector<vector<char>>& grid) {
        check[r][c] = true;
        
        for(int i=0; i<4; i++) {
            int nr = r+dr[i];
            int nc = c+dc[i];
            
            if(nr<0 || nc<0 || nr >= grid.size() || nc >= grid[0].size()) continue;
            
            if(!check[nr][nc] && grid[nr][nc] == '1') {
                dfs(nr,nc,grid);
            }
        }
    }
    
    
    int numIslands(vector<vector<char>>& grid) {
        int answer = 0;
        
        
        for(int i=0; i<grid.size(); i++) {
            for(int j=0; j<grid[0].size(); j++) {
                if(!check[i][j] && grid[i][j] == '1') {
                    answer++;
                    dfs(i,j,grid);
                }
            }
        }
        
        return answer;
    }
};

File: Leetcode_200. Number of Islands.py
# Runtime: 518 ms, faster than 43.20% of Python3 online submissions for Number of Islands.
# Memory Usage: 22.8 MB, less than 13.77% of Python3 online submissions for Number of Islands.

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        dx = [0,0,1,-1]
        dy = [1,-1,0,0]
        TMP = 301
        check = [[0]*TMP for _ in range(TMP) ]
        answer = 0
        
        def dfs(nowX, nowY):
            for n in range(4):
                nx = nowX + dx[n]
                ny = nowY + dy[n]
                if nx < 0 or ny < 0 or nx >= len(grid) or ny >=len(grid[0]):
                    continue
                if grid[nx][ny] == '1' and check[nx][ny] == 0:
                    check[nx][ny] = 1
                    dfs(nx,ny)
        
        for x in range(len(grid)):
            for y in range(len(grid[0])):
                if check[x][y] == 0 and grid[x][y] == '1':
                    check[x][y] = 1
                    answer += 1
                    dfs(x,y)
        
        return answer
            
